core/lib/dal/src/protocol_versions_dal.rs

```diff
@@ -98,6 +98,47 @@ impl ProtocolVersionsDal<'_, '_> {
         db_transaction.commit().await.unwrap();
     }
 
+    async fn save_genesis_upgrade(&mut self, id: ProtocolVersionId, tx_hash: Option<H256>) {
+        sqlx::query!(
+            r#"
+            UPDATE protocol_versions
+            SET
+                upgrade_tx_hash = $1
+            WHERE
+                id = $2
+            "#,
+            tx_hash.map(|tx_hash| tx_hash.0.to_vec()),
+            id as i32,
+        )
+        .execute(self.storage.conn())
+        .await
+        .unwrap();
+    }
+
+    /// Attaches a transaction used to set ChainId to the genesis protocol version.
+    /// Also inserts that transaction into the database.
+    pub async fn save_genesis_upgrade_with_tx(
+        &mut self,
+        id: ProtocolVersionId,
+        tx: ProtocolUpgradeTx,
+    ) {
+        let tx_hash = Some(tx.common_data.hash());
+
+        let mut db_transaction = self.storage.start_transaction().await.unwrap();
+
+        db_transaction
+            .transactions_dal()
+            .insert_system_transaction(tx)
+            .await;
```

Comment:
why do we combine this on the dal level? Dal shouldn't have any business logic. The fact that you need to update two tables is better represented on the app layer (`eth-watch` in this case). So please create the transaction there and call two dal methods from there

Reply:
it's the same as [`save_protocol_version_with_tx`](https://github.com/matter-labs/zksync-era/blob/689d38c3aea4d63da259362f9f96c51c3a359191/core/lib/dal/src/protocol_versions_dal.rs#L75) though?

Reply:
I think both have to be reworked, but there is no need to do it now. I'm resolving the convo.